### PR TITLE
fix(selectFirst): map a matched undefined element

### LINKED ISSUE #483: selectFirst skips the mapper when the matched element is undefined
`selectFirst` already tracks `foundIndex`, then still keys off the found *value*:

```ts
const found = array.find((item, index) => {
  foundIndex = index
  return condition ? condition(item, index) : mapper(item, index) != null
})
return found === undefined ? undefined : mapper(found, foundIndex)
```

`Array.prototype.find` returns `undefined` both when nothing matched and when the matching element *is* `undefined`. In the second case the mapper never runs.

Without a condition, the docs say it selects the first mapped value that is non-nullish. A mapper that produces a non-nullish result from an `undefined` element is still skipped.

### Repro (current main / 12.9.4)

```ts
import { selectFirst } from 'radashi'

selectFirst(
  [undefined, 1],
  x => (x === undefined ? 'none' : x),
  () => true,
)
// actual:   undefined
// expected: 'none'

selectFirst(
  [undefined, 1],
  x => (x === undefined ? 'none' : x),
)
// same result — mapper is never applied to the matched undefined element
```

### Suggested fix

Treat `foundIndex !== -1` as the hit, not `found !== undefined`. Initialize `foundIndex` to `-1` (already the case) and only skip the mapper when nothing matched.

Happy to send a PR if that's useful.

### PR DESCRIPTION
**Pull request by ethanstoner**
## Summary

`selectFirst` used `array.find(...)` and then checked `found === undefined` to decide whether anything matched. `find` returns `undefined` both when nothing matches and when the matched element is itself `undefined`, so in the second case the mapper was never called:

```ts
const mapper = (x: number | undefined) => (x === undefined ? 'none' : x)
selectFirst([undefined, 1], mapper, () => true) // was undefined, now 'none'
selectFirst([undefined, 1], mapper)             // was undefined, now 'none'
```

The fix uses `findIndex` and checks `foundIndex === -1`, so the decision is based on whether an index matched, not on the element's value. Signature, types, and all other behavior are unchanged. A regression test covers both the explicit-condition and default-condition cases.

This change was made with the help of Claude Code; I reviewed and tested it.

## Related issue, if any:

Fixes #483

## For any code change,

- [ ] Related documentation has been updated, if needed
- [x] Related tests have been added or updated, if needed
- [ ] Related benchmarks have been added or updated, if needed
- [ ] Release notes in [next-minor.md](.github/next-minor.md) or [next-major.md](.github/next-major.md) have been added, if needed

## Does this PR introduce a breaking change?

No

## Bundle impact

| Status | File | Size [^1337] | Difference |
| --- | --- | --- | --- |
| M | `src/array/selectFirst.ts` | 141 | -2 (-1%) |

[^1337]: Function size includes the `import` dependencies of the function.

### CODE REVIEW

**Comment by MarlonPassos-git** [comment]
Thanks for the PR! I was taking a look at this and thought of a pretty weird edge case 😅

I'm not sure how relevant this is in practice, but there is a behavior change here.

For example, imagine we had this test:

```ts
it('maps the same value that matched the condition', () => {
  const values = [1]

  const result = selectFirst(
    values,
    value => value * 10,
    () => {
      values[0] = 2
      return true
    },
  )

  expect(result).toBe(10)
})
```

With the current implementation, this passes because `find` keeps the value that actually matched (`1`).

With this PR, the test would fail and return `20`, because we first find the index and then read the array again here:

```ts
mapper(array[foundIndex], foundIndex)
```

At that point, `array[0]` is already `2`.

I know this is a pretty unusual edge case and we don't currently test this behavior, but it does look like a regression from the previous implementation.

Maybe we should either explicitly accept and test this new behavior, or adjust the implementation so the mapper still receives the same value that originally matched the condition.

Personally, I'd prefer preserving the previous behavior.

File: tests/array/selectFirst.test.ts [hunk 1]
```diff
@@ -57,4 +57,10 @@ describe('selectFirst', () => {
     const result = _.selectFirst(list, el => el.a)
     expect(result).toEqual(1)
   })
+  test('maps a matched undefined element', () => {
```
Comment (MarlonPassos-git) [inline tests/array/selectFirst.test.ts:60]: Maybe worth adding one more case where the matched `undefined` is not at index `0`, just to make sure we also cover the index/short-circuit behavior here.

Something like:

```ts
const values = [0, undefined, 2]

expect(
  selectFirst(
    values,
    value => (value === undefined ? 'none' : value),
    (_, index) => index === 1,
  ),
).toBe('none')
```

Comment (ethanstoner) [reply tests/array/selectFirst.test.ts:60]: Added, as `maps a matched undefined element past the first index` — your exact case (`[0, undefined, 2]` matching on `index === 1`). It confirms the short-circuit picks the right index rather than just the first element.

**Comment by ethanstoner** [comment]
You're right, and it's a real regression rather than a theoretical one — I reproduced it with your test verbatim: on `fee9452^` it fails with `expected 20 to deeply equal 10`.

The cause is exactly as you described. `find` handed the mapper the value captured at match time, whereas `array[foundIndex]` re-reads the array after `findIndex` has returned, so a mutating condition changes what the mapper sees.

I went with preserving the previous behavior, since restoring `find` semantics keeps this a pure bug fix for #483 with no other observable change. The scan now captures the matched item:

```ts
let foundItem!: T
const foundIndex = array.findIndex((item, index) => {
  foundItem = item
  return condition ? condition(item, index) : mapper(item, index) != null
})
return foundIndex === -1 ? undefined : mapper(foundItem, foundIndex)
```

Your test is committed as `maps the same value that matched the condition`, so the behavior is now pinned rather than merely untested. I also added the index case you suggested inline.

Verified: `182` test files / `914` tests pass, no type errors, biome clean, and `selectFirst.ts` stays at 100% coverage on all four metrics.
